object_store/src/lib.rs

```diff
@@ -848,6 +1053,263 @@ mod tests {
 
             r
         }
+
+        async fn get_nonexistant_object(
+            storage: &ObjectStore,
+            location: Option<&str>,
+        ) -> Result<Bytes> {
+            let location = location.unwrap_or("this_file_should_not_exist");
+
+            let content_list = flatten_list_stream(storage, Some(location)).await?;
+            assert!(content_list.is_empty());
+
+            Ok(storage
+                .get(location)
+                .await?
+                .map_ok(|b| bytes::BytesMut::from(&b[..]))
+                .try_concat()
+                .await?
+                .freeze())
+        }
+
+        #[tokio::test]
+        async fn s3_test() -> Result<()> {
+            let (region, bucket_name) = region_and_bucket_name()?;
+
+            let integration = ObjectStore::new_amazon_s3(AmazonS3::new(region, &bucket_name));
+            check_credentials(put_get_delete_list(&integration).await)?;
+
+            Ok(())
+        }
+
+        #[tokio::test]
+        async fn s3_test_get_nonexistant_region() -> Result<()> {
+            // Assumes environment variables do not provide credentials to AWS US West 1
+            let (_, bucket_name) = region_and_bucket_name()?;
+            let region = rusoto_core::Region::UsWest1;
+            let integration = ObjectStore::new_amazon_s3(AmazonS3::new(region, &bucket_name));
+            let location_name = NON_EXISTANT_NAME;
+
+            let result = get_nonexistant_object(&integration, Some(location_name)).await;
+            if let Err(e) = result {
+                if let Some(Error(InternalError::UnableToListDataFromS3 { source, bucket })) =
+                    e.downcast_ref::<crate::Error>()
+                {
+                    assert!(matches!(source, rusoto_core::RusotoError::Unknown(_)));
+                    assert_eq!(bucket, &bucket_name);
+                } else {
+                    panic!()
+                }
+            } else {
+                panic!()
+            }
```

Comment:
We could unwrap the error to remove one level of `if let` unwrapping :)

```suggestion
            // Unwrap the error - this will panic if it is an Ok(_)
            let err = get_nonexistant_object(&integration, Some(location_name))
                .await
                .unwrap_err();

            // Unwrap & Destructure the error
            if let Error(InternalError::UnableToListDataFromS3 { source, bucket }) =
                err.downcast_ref::<crate::Error>().unwrap()
            {
                // Assert!
                assert!(matches!(source, rusoto_core::RusotoError::Unknown(_)));
                assert_eq!(bucket, &bucket_name);
            } else {
                panic!("unexpected error type")
            }
```

Reply:
:man_facepalming: Thanks! Simplified the error checking in the tests to this pattern.